### PR TITLE
Add CSP and UISecurity directives and keywords

### LINKED ISSUE #2663: Incomplete support for CSP
**Information**

- Language: CSP
- Plugins: none

**Does the problem still occur in the latest version of Prism?**

[Yes](https://prismjs.com/test.html#language=csp).

**Description**

**_New directives and keywords_**

From the latest (2018-10-15) [working draft](https://www.w3.org/TR/2018/WD-CSP3-20181015/) of CSP3:

- `navigate-to`;
- `prefetch-src`;
- `script-src-attr`;
- `script-src-elem`;
- `style-src-attr`;
- `style-src-elem`;
- `'report-sample'`;
- `'unsafe-allow-redirects'`;
- `'unsafe-hashes'`.

From [Trusted Types](https://w3c.github.io/webappsec-trusted-types/dist/spec/):

- `require-trusted-types-for`;
- `trusted-types`;
- `'allow-duplicates'`;
- `'script'`.

These are partially implemented in Blink (Chromium, Chrome, etc.) and Gecko (Firefox).

Also, Google Cobalt [recognizes](https://github.com/youtube/cobalt/blob/0d1858f87ad39776d3bccd0ec75731daf232e0d3/src/cobalt/csp/content_security_policy.cc#L155-L157) the [`suborigin`](https://metromoxie.github.io/webappsec/specs/suborigins/#the-suborigin-directive) directive.

By the way, there is a bunch of relevant (`input-protection`) and dropped (`input-protection-clip`, `input-protection-selectors`) directives from [User Interface Security and the Visibility API](https://www.w3.org/TR/2016/WD-UISecurity-20160607/), but I don’t know of any software that uses these directives.

**_Old directives and keywords_**

Prism supports some directives and keywords, which are dropped from CSP (see the table below).

|Token                       |Introduced in                                                           |Removed from                                                        |
|----------------------------|------------------------------------------------------------------------|--------------------------------------------------------------------|
|`disown-opener`             |[CSP3. WD. 2016-08-01](https://www.w3.org/TR/2016/WD-CSP3-20160801/)    |[CSP3. WD. 2018-10-15](https://www.w3.org/TR/2018/WD-CSP3-20181015/)|
|`referrer`                  |[CSP 1.1. WD. 2014-02-11](https://www.w3.org/TR/2014/WD-CSP11-20140211/)|[CSP2. CR. 2015-02-19](https://www.w3.org/TR/2015/CR-CSP2-20150219/)|
|`reflected-xss`             |[CSP 1.1. WD. 2012-12-13](https://www.w3.org/TR/2012/WD-CSP11-20121213/)|[CSP2. CR. 2015-02-19](https://www.w3.org/TR/2015/CR-CSP2-20150219/)|
|`'unsafe-hashed-attributes'`|[CSP3. WD. 2016-06-21](https://www.w3.org/TR/2016/WD-CSP3-20160621/)    |[CSP3. WD. 2018-10-15](https://www.w3.org/TR/2018/WD-CSP3-20181015/)|

[Gecko](https://github.com/mozilla/gecko-dev/blob/cd8b8939b911b3c28c13de23fde7201a8452499b/dom/security/nsCSPParser.cpp#L861-L871) still recognizes `reflected-xss` as a directive, albeit ignored.

Here are some directives and keywords, which are also dropped from CSP and have no support in Prism (see the table below).

|Token                     |Introduced in                                                         |Removed from                                                          |
|--------------------------|----------------------------------------------------------------------|----------------------------------------------------------------------|
|`policy-uri`              |[CSP 1.0. WD. 2011-11-29](https://www.w3.org/TR/2011/WD-CSP-20111129/)|[CSP 1.0. WD. 2012-07-10](https://www.w3.org/TR/2012/WD-CSP-20120710/)|
|`'unsafe-hash-attributes'`|[CSP3. WD. 2016-04-25](https://www.w3.org/TR/2016/WD-CSP3-20160425/)  |[CSP3. WD. 2016-06-21](https://www.w3.org/TR/2016/WD-CSP3-20160621/)  |

[Blink](https://github.com/chromium/chromium/blob/004ed837b5ea0e08cc209f36b9f274f43a51d439/third_party/blink/renderer/core/frame/csp/content_security_policy.cc#L1294-L1297) recognizes `policy-uri` as a “removed from the specification” directive. Should `policy-uri` and `'unsafe-hash-attributes'` be added for consistency?

**Code snippet**

<details>

<summary>The code being highlighted incorrectly.</summary>

```
default-src 'report-sample' 'unsafe-hash-attributes' 'unsafe-hashes';
navigate-to 'unsafe-allow-redirects';
prefetch-src;
script-src-attr;
script-src-elem;
style-src-attr;
style-src-elem;

require-trusted-types-for 'script';
trusted-types foo bar 'allow-duplicates';

suborigin;

input-protection;
input-protection-clip;
input-protection-selectors;

policy-uri https://example.com;
```

<del>(`script-src-attr`, `script-src-elem`, `style-src-attr`, `style-src-elem` are partly highlighted due to #2661.)</del>

</details>

**Questions**

1. Should I add new directives and keywords from CSP and Trusted Types?
2. Should I add `suborigin`?
3. Should I add `input-protection`?
4. Should I add `input-protection-clip` and `input-protection-selectors`?
5. Should I **remove** directives and keywords, which are dropped from CSP (`disown-opener`, `referrer`, `reflected-xss`, `'unsafe-hashed-attributes'`)?
6. Should I **add** directives and keywords, which are dropped from CSP (`policy-uri` and `'unsafe-hash-attributes'`)?

### PR DESCRIPTION
**Pull request by edukisto**
Resolves #2663.

### CODE REVIEW

File: components/prism-csp.js [hunk 1]
```diff
@@ -10,17 +10,17 @@
...
-	'directive':  {
-		pattern: /(^|[^-\da-z])(?:base-uri|block-all-mixed-content|(?:child|connect|default|font|frame|img|manifest|media|object|script|style|worker)-src|disown-opener|form-action|frame-ancestors|plugin-types|referrer|reflected-xss|report-to|report-uri|require-sri-for|sandbox|upgrade-insecure-requests)(?=[^-\da-z]|$)/i,
+	'directive': {
+		pattern: /(^|[^-\da-z])(?:base-uri|block-all-mixed-content|(?:child|connect|default|font|frame|img|manifest|media|object|prefetch|script|style|worker)-src|disown-opener|form-action|frame-(?:ancestors|options)|input-protection(?:-(?:clip|selectors))?|navigate-to|plugin-types|policy-uri|referrer|reflected-xss|report-(?:to|uri)|require-sri-for|sandbox|(?:script|style)-src-(?:attr|elem)|upgrade-insecure-requests)(?=[^-\da-z]|$)/i,
 		lookbehind: true,
 		alias: 'keyword'
 	},
 	'safe': {
-		pattern: /'(?:self|none|strict-dynamic|(?:nonce-|sha(?:256|384|512)-)[a-zA-Z\d+=/]+)'/,
+		pattern: /'(?:deny|none|report-sample|self|strict-dynamic|top-only|(?:nonce|sha(?:256|384|512))-[-+/\d=_a-z]+)'/i,
```
Comment (RunDevelopment) [inline components/prism-csp.js:19]: Could you please add a comment with a link to the base64 spec you linked in #2663? Without that document, `[-+/\d=_a-z]` is hard to make sense of.

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
One nit but otherwise good to go.

File: components/prism-csp.js (hunk 1 above)
Comment (edukisto) [reply components/prism-csp.js:22]: Done!

**Review by RunDevelopment — APPROVED** [review]

**Comment by RunDevelopment** [comment]
Thank you for contributing @edukisto!

**Comment by edukisto** [comment]
Thank you, @RunDevelopment!

**Pull request merged**